Add basic TypeScript documentation (#144)
WoltLab/woltlab.github.io: docs/javascript/typescript.md

# TypeScript

## Consuming WoltLab Suite’s Types

> To consume the types of WoltLab Suite you will need to install the [WoltLab/WCF](https://github.com/WoltLab/WCF) using npm using a git URL that refers to the appropriate branch.

Reviewer: ```suggestion
To consume the types of WoltLab Suite, you will need to install the [WoltLab/WCF](https://github.com/WoltLab/WCF) using npm using a git URL that refers to the appropriate branch.
```

"… install the WoltLab/WCF *package*"?

Also using "using" twice: May-be "via npm using …"?

Author: I've pushed a fix. PTAL.

A full `package.json` that includes WoltLab Suite, TypeScript, eslint and Prettier could look like the following.

```json
{
  "devDependencies": {
    "@typescript-eslint/eslint-plugin": "^4.6.1",
    "@typescript-eslint/parser": "^4.6.1",
    "eslint": "^7.12.1",
    "eslint-config-prettier": "^6.15.0",
    "prettier": "^2.1.2",
    "tslib": "^2.0.3",
    "typescript": "^4.1.3"
  },
  "dependencies": {
    "@woltlab/wcf": "https://github.com/WoltLab/WCF.git#master"
  }
}
```

> After installing the types using npm you will also need to configure tsconfig.json to take the types into account.

Reviewer: ```suggestion
After installing the types using npm, you will also need to configure `tsconfig.json` to take the types into account.
```

May-be add the relevant `npm` commands (might apply in other places).

Author: I don't think we should explain how npm works, it's broadly known and we are not doing anything fancy. It's a simple `npm install`. I've only explained the use of `tsc` due to the use of `npx` in there, which is somewhat lesser known.

> To do so you will need to add them to the `compilerOptions.paths` option.

Reviewer: ```suggestion
To do so, you will need to add them to the `compilerOptions.paths` option.
```

> A full `tsconfig.json` that matches the configuration of WoltLab Suite could look like the following.

Reviewer: ```suggestion
A complete `tsconfig.json` file that matches the configuration of WoltLab Suite could look like the following.
```

```json
{
  "include": [
    "node_modules/@woltlab/wcf/global.d.ts",
    "ts/**/*"
  ],
  "compilerOptions": {
    "target": "es2017",
    "module": "amd",
    "rootDir": "ts/",
    "outDir": "files/js/",
    "lib": [
      "dom",
      "es2017"
    ],
    "strictNullChecks": true,
    "moduleResolution": "node",
    "esModuleInterop": true,
    "noImplicitThis": true,
    "strictBindCallApply": true,
    "baseUrl": ".",
    "paths": {
      "*": [
        "node_modules/@woltlab/wcf/ts/*"
      ]
    },
    "importHelpers": true
  }
}
```

> After this initial set-up you would place your TypeScript source files into the `ts/` folder of your project.

Reviewer: ```suggestion
After this initial set-up, you would place your TypeScript source files into the `ts/` folder of your project.
```

```

## Writing a simple module

After completing this initial set-up you can start writing your first TypeScript module.
The TypeScript compiler can be launched in Watch Mode by running `npx tsc -w`.

WoltLab Suite’s modules can be imported using the standard ECMAScript module import syntax by specifying the full module name.
The public API of the module can also be exported using the standard ECMASCript module export syntax.

```typescript
import * as Language from "WoltLabSuite/Core/Language";

export function run() {
  alert(Language.get("wcf.foo.bar"));
}
```

This simple example module will compile to plain JavaScript that is compatible with the AMD loader that is used by WoltLab Suite.

```javascript
define(["require", "exports", "tslib", "WoltLabSuite/Core/Language"], function (require, exports, tslib_1, Language) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.run = void 0;
    Language = tslib_1.__importStar(Language);
    function run() {
        alert(Language.get("wcf.foo.bar"));
    }
    exports.run = run;
});
```

Within templates it can be consumed as follows.

```html
<script data-relocate="true">

>   require(["Example"], function (Example) {

Reviewer: ```suggestion
  require(["Example"], (Example) => {
```
